Hm looks like the failing tests were due to Unpack not existing pre python3.11. I was able to run the tests successfully locally:

```
~/Develop/python-firestore (ig/add-types) $ nox -s lint
nox > Running session lint
nox > Creating virtual environment (virtualenv) using python3.14 in .nox/lint
nox > python -m pip install flake8==6.1.0 'black[jupyter]==23.7.0'
nox > black --check docs google tests noxfile.py setup.py
All done! ✨ 🍰 ✨
150 files would be left unchanged.
nox > flake8 google tests
nox > Session lint was successful in 6 seconds.
```

```
nox > py.test --quiet --junitxml=unit_3.14_sponge_log.xml --cov=google --cov=tests/unit --cov-append --cov-config=.coveragerc --cov-report= --cov-fail-under=0 tests/unit
......................................................................................................................... [  4%]
......................................................................................................................... [  8%]
......................................................................................................................... [ 12%]
......................................................................................................................... [ 16%]
......................................................................................................................... [ 20%]
......................................................................................................................... [ 24%]
......................................................................................................................... [ 28%]
......................................................................................................................... [ 32%]
......................................................................................................................... [ 36%]
......................................................................................................................... [ 40%]
......................................................................................................................... [ 44%]
......................................................................................................................... [ 48%]
......................................................................................................................... [ 52%]
......................................................................................................................... [ 56%]
......................................................................................................................... [ 60%]
......................................................................................................................... [ 64%]
..................................................................X...................................................... [ 68%]
......................................................................................................................... [ 72%]
......................................................................................................................... [ 76%]
......................................................................................................................... [ 80%]
..................................................................................sssssssssssssssssssssssssssssssssssssss [ 84%]
sssss.................................................................................................................... [ 88%]
......................................................................................................................... [ 92%]
......................................................................................................................... [ 96%]
...................................................................................................................       [100%]
2974 passed, 44 skipped, 1 xpassed in 19.27s
nox > Session unit-3.14(protobuf_implementation='python') was successful in 29 seconds.
```

The one weird thing I see is this line:
```
.................................................................................................Thread-ConsumeBidirectionalStream caught unexpected exception 'Mock' object is not an iterator and will exit.
Traceback (most recent call last):
  File "/Users/ishaangandhi/Develop/python-firestore/.nox/unit-3-12-protobuf_implementation-upb/lib/python3.12/site-packages/google/api_core/bidi.py", line 644, in _thread_main
    response = self._bidi_rpc.recv()
               ^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ishaangandhi/Develop/python-firestore/.nox/unit-3-12-protobuf_implementation-upb/lib/python3.12/site-packages/google/api_core/bidi.py", line 547, in recv
    return self._recoverable(self._recv)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ishaangandhi/Develop/python-firestore/.nox/unit-3-12-protobuf_implementation-upb/lib/python3.12/site-packages/google/api_core/bidi.py", line 506, in _recoverable
    raise exc
  File "/Users/ishaangandhi/Develop/python-firestore/.nox/unit-3-12-protobuf_implementation-upb/lib/python3.12/site-packages/google/api_core/bidi.py", line 490, in _recoverable
    return method(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ishaangandhi/Develop/python-firestore/.nox/unit-3-12-protobuf_implementation-upb/lib/python3.12/site-packages/google/api_core/bidi.py", line 544, in _recv
    return next(call)
           ^^^^^^^^^^
TypeError: 'Mock' object is not an iterator
```

But it still says the tests passed, and the traceback only goes over lines within `.nox/unit-3-12-protobuf_implementation-upb`.

--- a/google/cloud/firestore_v1/async_batch.py
+++ b/google/cloud/firestore_v1/async_batch.py
@@ -19,6 +19,7 @@ from google.api_core import gapic_v1
 from google.api_core import retry_async as retries
 
 from google.cloud.firestore_v1.base_batch import BaseWriteBatch
+from google.cloud.firestore_v1.types.write import WriteResult
 
 
 class AsyncWriteBatch(BaseWriteBatch):
@@ -40,7 +41,7 @@ class AsyncWriteBatch(BaseWriteBatch):
         self,
         retry: retries.AsyncRetry | object | None = gapic_v1.method.DEFAULT,
         timeout: float | None = None,
-    ) -> list:
+    ) -> list[WriteResult]:
         """Commit the changes accumulated in this batch.
 
         Args:

--- a/google/cloud/firestore_v1/async_client.py
+++ b/google/cloud/firestore_v1/async_client.py
@@ -25,7 +25,15 @@ In the hierarchy of API concepts
 """
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Any, AsyncGenerator, Iterable, List, Optional, Union
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    AsyncGenerator,
+    Iterable,
+    List,
+    Optional,
+    Union,
+)
 
 from google.api_core import gapic_v1
 from google.api_core import retry_async as retries
@@ -40,6 +48,7 @@ from google.cloud.firestore_v1.async_query import AsyncCollectionGroup
 from google.cloud.firestore_v1.async_transaction import AsyncTransaction
 from google.cloud.firestore_v1.base_client import _parse_batch_get  # type: ignore
 from google.cloud.firestore_v1.base_client import _CLIENT_INFO, BaseClient, _path_helper
+from google.cloud.firestore_v1.base_transaction import MAX_ATTEMPTS
 from google.cloud.firestore_v1.field_path import FieldPath
 from google.cloud.firestore_v1.services.firestore import (
     async_client as firestore_client,
@@ -410,7 +419,9 @@ class AsyncClient(BaseClient):
         """
         return AsyncWriteBatch(self)
 
-    def transaction(self, **kwargs) -> AsyncTransaction:
+    def transaction(
+        self, max_attempts: int = MAX_ATTEMPTS, read_only: bool = False
+    ) -> AsyncTransaction:
         """Get a transaction that uses this client.
 
         See :class:`~google.cloud.firestore_v1.async_transaction.AsyncTransaction` for
@@ -426,4 +437,4 @@ class AsyncClient(BaseClient):
             :class:`~google.cloud.firestore_v1.async_transaction.AsyncTransaction`:
             A transaction attached to this client.
         """
-        return AsyncTransaction(self, **kwargs)
+        return AsyncTransaction(self, max_attempts=max_attempts, read_only=read_only)

--- a/google/cloud/firestore_v1/async_collection.py
+++ b/google/cloud/firestore_v1/async_collection.py
@@ -15,7 +15,7 @@
 """Classes for representing collections for the Google Cloud Firestore API."""
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Any, AsyncGenerator, Optional, Tuple
+from typing import TYPE_CHECKING, Any, AsyncGenerator, Optional, Tuple, cast
 
 from google.api_core import gapic_v1
 from google.api_core import retry_async as retries
@@ -153,7 +153,8 @@ class AsyncCollectionReference(BaseCollectionReference[async_query.AsyncQuery]):
             :class:`~google.cloud.firestore_v1.document.async_document.AsyncDocumentReference`:
             The child document.
         """
-        return super(AsyncCollectionReference, self).document(document_id)
+        doc = super(AsyncCollectionReference, self).document(document_id)
+        return cast("AsyncDocumentReference", doc)
 
     async def list_documents(
         self,

--- a/google/cloud/firestore_v1/base_batch.py
+++ b/google/cloud/firestore_v1/base_batch.py
@@ -15,7 +15,7 @@
 """Helpers for batch requests to the Google Cloud Firestore API."""
 from __future__ import annotations
 import abc
-from typing import Dict, Union
+from typing import Any, Dict, Union
 
 # Types needed only for Type Hints
 from google.api_core import retry as retries
@@ -67,7 +67,9 @@ class BaseBatch(metaclass=abc.ABCMeta):
         write depend on the implementing class."""
         raise NotImplementedError()
 
-    def create(self, reference: BaseDocumentReference, document_data: dict) -> None:
+    def create(
+        self, reference: BaseDocumentReference, document_data: dict[str, Any]
+    ) -> None:
         """Add a "change" to this batch to create a document.
 
         If the document given by ``reference`` already exists, then this
@@ -120,7 +122,7 @@ class BaseBatch(metaclass=abc.ABCMeta):
     def update(
         self,
         reference: BaseDocumentReference,
-        field_updates: dict,
+        field_updates: dict[str, Any],
         option: _helpers.WriteOption | None = None,
     ) -> None:
         """Add a "change" to update a document.

--- a/google/cloud/firestore_v1/base_client.py
+++ b/google/cloud/firestore_v1/base_client.py
@@ -57,7 +57,7 @@ from google.cloud.firestore_v1.base_document import (
     DocumentSnapshot,
 )
 from google.cloud.firestore_v1.base_query import BaseQuery
-from google.cloud.firestore_v1.base_transaction import BaseTransaction
+from google.cloud.firestore_v1.base_transaction import MAX_ATTEMPTS, BaseTransaction
 from google.cloud.firestore_v1.bulk_writer import BulkWriter, BulkWriterOptions
 from google.cloud.firestore_v1.field_path import render_field_path
 from google.cloud.firestore_v1.services.firestore import client as firestore_client
@@ -497,7 +497,9 @@ class BaseClient(ClientWithProject):
     def batch(self) -> BaseWriteBatch:
         raise NotImplementedError
 
-    def transaction(self, **kwargs) -> BaseTransaction:
+    def transaction(
+        self, max_attempts: int = MAX_ATTEMPTS, read_only: bool = False
+    ) -> BaseTransaction:
         raise NotImplementedError
 
 

--- a/google/cloud/firestore_v1/base_collection.py
+++ b/google/cloud/firestore_v1/base_collection.py
@@ -35,6 +35,7 @@ from typing import (
 from google.api_core import retry as retries
 
 from google.cloud.firestore_v1 import _helpers
+from google.cloud.firestore_v1.base_document import BaseDocumentReference
 from google.cloud.firestore_v1.base_query import QueryType
 
 if TYPE_CHECKING:  # pragma: NO COVER
@@ -133,7 +134,7 @@ class BaseCollectionReference(Generic[QueryType]):
     def _vector_query(self) -> BaseVectorQuery:
         raise NotImplementedError
 
-    def document(self, document_id: Optional[str] = None):
+    def document(self, document_id: Optional[str] = None) -> BaseDocumentReference:
         """Create a sub-document underneath the current collection.
 
         Args:

--- a/google/cloud/firestore_v1/base_document.py
+++ b/google/cloud/firestore_v1/base_document.py
@@ -418,7 +418,7 @@ class DocumentSnapshot(object):
         return self._reference._client
 
     @property
-    def exists(self):
+    def exists(self) -> bool:
         """Existence flag.
 
         Indicates if the document existed at the time this snapshot
@@ -430,7 +430,7 @@ class DocumentSnapshot(object):
         return self._exists
 
     @property
-    def id(self):
+    def id(self) -> str:
         """The document identifier (within its collection).
 
         Returns:
@@ -439,7 +439,7 @@ class DocumentSnapshot(object):
         return self._reference.id
 
     @property
-    def reference(self):
+    def reference(self) -> BaseDocumentReference:
         """Document reference corresponding to document that owns this data.
 
         Returns:

--- a/google/cloud/firestore_v1/client.py
+++ b/google/cloud/firestore_v1/client.py
@@ -39,6 +39,7 @@ from google.cloud.firestore_v1.base_client import (
 
 # Types needed only for Type Hints
 from google.cloud.firestore_v1.base_document import DocumentSnapshot
+from google.cloud.firestore_v1.base_transaction import MAX_ATTEMPTS
 from google.cloud.firestore_v1.batch import WriteBatch
 from google.cloud.firestore_v1.collection import CollectionReference
 from google.cloud.firestore_v1.document import DocumentReference
@@ -391,7 +392,9 @@ class Client(BaseClient):
         """
         return WriteBatch(self)
 
-    def transaction(self, **kwargs) -> Transaction:
+    def transaction(
+        self, max_attempts: int = MAX_ATTEMPTS, read_only: bool = False
+    ) -> Transaction:
         """Get a transaction that uses this client.
 
         See :class:`~google.cloud.firestore_v1.transaction.Transaction` for
@@ -407,4 +410,4 @@ class Client(BaseClient):
             :class:`~google.cloud.firestore_v1.transaction.Transaction`:
             A transaction attached to this client.
         """
-        return Transaction(self, **kwargs)
+        return Transaction(self, max_attempts=max_attempts, read_only=read_only)

--- a/google/cloud/firestore_v1/document.py
+++ b/google/cloud/firestore_v1/document.py
@@ -169,7 +169,7 @@ class DocumentReference(BaseDocumentReference):
 
     def update(
         self,
-        field_updates: dict,
+        field_updates: dict[str, Any],
         option: _helpers.WriteOption | None = None,
         retry: retries.Retry | object | None = gapic_v1.method.DEFAULT,
         timeout: float | None = None,

